Fix crossgen2 to fix some negative tests

janvorli opened the pull request:
This change fixes 4 negative tests in the Loader\classloader\generics\Instantiation\Negative 
folder.

The tests were trying to instantiate objects using generic definition types
without supplying generic arguments.

janvorli committed:
Fix crossgen2 to fix remaining negative tests

This change fixes about 12 negative tests in the
Loader\classloader\generics\Instantiation\Negative folder and several
others.

In most cases, the tests were trying to instantiate objects using
generic definition types without supplying generic arguments,
instantiating types derived from abstract classes without implementation
of one of the methods, similar instantiating types implementing
interface but not implementing one of the methods or invoking methods on
generic definition types.

janvorli committed:
Reflect PR feedback

janvorli commented:
This change was already reviewed by @davidwrighton, but I had to recreate the PR due to the switch of the runtime repo from private to public.

MichalStrehovsky commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -1357,6 +1357,11 @@ private uint getClassSize(CORINFO_CLASS_STRUCT_* cls)
                 throw new RequiresRuntimeJitException(type);
             }
 #endif
+            if (type.IsGenericDefinition)
```
Would be better to prevent computing the layout for uninstantiated types instead (so that it throws in `GetElementSize` a couple lines above). Adding the check to `MetadataFieldLayoutAlgorithm.cs` might be better - I would place it somewhere where we do the other `ThrowHelper.ThrowTypeLoadException` in `ComputeInstanceLayout`.

(Also note this is TypeLayoutException, not InvalidProgramException. It will matter more when/if we start pregenerating throwing method bodies for these - it already matters for CoreRT.)

janvorli replied:
Thank you for the suggestion, I'll give it a try.

trylek reviewed (approved):
LGTM modulo Michal's suggestion - if that works, it would probably make the change cleaner and simpler.

janvorli committed:
Reflect PR feedback

Move the uninstantiated type check to MetadataFieldLayoutAlgorithm.ComputeInstanceLayout
and change the exception to TypeLoadException.

janvorli committed:
Remove check for IsGenericDefinition from getNewHelper

With the previous change, it is not needed anymore.

janvorli commented:
I've also removed the check from the `getNewHelper`, with the change @MichalStrehovsky has suggested, it is not needed anymore as the check in `ComputeInstanceLayout` covers that case too.

MichalStrehovsky commented on src/coreclr/src/tools/Common/TypeSystem/Common/ExceptionStringID.cs:
```diff
@@ -16,6 +16,7 @@ public enum ExceptionStringID
         ClassLoadExplicitLayout,
         ClassLoadValueClassTooLarge,
         ClassLoadRankTooLarge,
+        ClassLoadUninstantiatedType,
```
In think we can get by with ClassLoadGeneral. CoreCLR also doesn't have a special resource string for this case.

MichalStrehovsky commented on src/coreclr/src/tools/Common/TypeSystem/Common/MetadataFieldLayoutAlgorithm.cs:
```diff
@@ -17,6 +17,12 @@ public class MetadataFieldLayoutAlgorithm : FieldLayoutAlgorithm
+
+            if (type.IsGenericDefinition)
+            {
+                ThrowHelper.ThrowTypeLoadException(ExceptionStringID.ClassLoadUninstantiatedType, type.GetTypeDefinition());
```
Nit: this already is a type definition. GetTypeDefinition is a no-op.

MichalStrehovsky reviewed (approved):
> I've also removed the check

Nice bonus!

janvorli committed:
Reflect PR feedback
